File: demos/showcase/android/compose/src/main/java/com/bajutsu/showcase/compose/GestureScreen.kt
package com.bajutsu.showcase.compose

import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectTransformGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.unit.dp

// BE-0232: the Compose realization of the two-finger gesture screen, mirroring the iOS GestureView
// (BE-0019). Pinch and rotate are the one class of actuation a single-touch backend (like the retired
// idb backend, BE-0290) cannot perform, and on adb they need a rooted `sendevent` sweep; each target flips its mirrored a11y value
// once its gesture is recognized, so the shared `gestures_multitouch` run can assert the actuation
// landed rather than merely not erroring. Reached only when the SHOWCASE_GESTURES launch env is set
// (see AppModel / RootScreen), so the normal five-tab app (BE-0079) never renders it. A flat, scroll-
// free Column so both targets are always in the tree — a two-finger target must be fully on-screen for
// its touch points anyway, and the shared scenario reads their ids (log.pinch / log.rotate) unchanged.
@Composable
fun GestureScreen() {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .enableTestTagsAsResourceId(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(24.dp, Alignment.CenterVertically),
    ) {
        GestureTarget("Pinch me", "log.pinch", pressedLabel = "pinched") { zoom, _ -> zoom != 1f }
        GestureTarget("Rotate me", "log.rotate", pressedLabel = "rotated") { _, rot -> rot != 0f }
    }
}

// A generous, opaque hit area (as on iOS): a pinch/rotate is two touch points that need real room,
// and the gesture degenerates on an intrinsically-sized view. `detectTransformGestures` reports pan,
// zoom, and rotation together; `recognized` picks the channel this target cares about (zoom for
// pinch, rotation for rotate), so a pure pinch does not trip the rotate target and vice versa. The id
// (log.pinch) tags the hit area and its `.value` twin mirrors the result into content-desc — the same
// aid + stateValue pairing the Log tab uses (SPEC §8), so the shared scenario reads it unchanged.
@Composable
private fun GestureTarget(
    label: String,
    id: String,
    pressedLabel: String,
    recognized: (zoom: Float, rotation: Float) -> Boolean,
) {
    var fired by remember { mutableStateOf(false) }
    Text(
        text = label,
        modifier = Modifier
            .size(width = 280.dp, height = 120.dp)
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .aid(id)
            .pointerInput(Unit) {
                detectTransformGestures { _, _, zoom, rotation ->
                    if (recognized(zoom, rotation)) fired = true
                }
            },
    )
    val value = if (fired) pressedLabel else "idle"
    Text(value, Modifier.aid("$id.value").stateValue(value))
}

File: demos/showcase/flutter/lib/screens/gesture_screen.dart
import 'package:flutter/material.dart';

import '../accessibility.dart';

/// BE-0232: the two-finger gesture screen, the Flutter twin of the iOS `GestureView` and the Compose
/// `GestureScreen`. Pinch and rotate are the one class of actuation a single-touch backend cannot
/// perform; each target flips its mirrored a11y value once its gesture is recognized, so the shared
/// `gestures_multitouch` run can assert the actuation landed rather than merely not erroring. Reached
/// only when `SHOWCASE_GESTURES` is set, so the normal five-tab app never renders it. A flat,
/// scroll-free column keeps both targets always on-screen (a two-finger target needs full room).
class GestureScreen extends StatelessWidget {
  const GestureScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: SafeArea(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            _GestureTarget(label: 'Pinch me', id: 'log.pinch', pressedLabel: 'pinched', channel: _Channel.zoom),
            SizedBox(height: 24),
            _GestureTarget(label: 'Rotate me', id: 'log.rotate', pressedLabel: 'rotated', channel: _Channel.rotation),
          ],
        ),
      ),
    );
  }
}

enum _Channel { zoom, rotation }

/// A generous, opaque hit area. `onScaleUpdate` reports scale and rotation together; [channel]
/// picks the one this target cares about, so a pure pinch does not trip the rotate target and vice
/// versa. The id tags the hit area and its `.value` twin mirrors the result — the same aid + value
/// pairing the Log tab uses (SPEC §8), so the shared scenario reads it unchanged.
class _GestureTarget extends StatefulWidget {
  const _GestureTarget({required this.label, required this.id, required this.pressedLabel, required this.channel});

  final String label;
  final String id;
  final String pressedLabel;
  final _Channel channel;

  @override
  State<_GestureTarget> createState() => _GestureTargetState();
}

class _GestureTargetState extends State<_GestureTarget> {
  bool _fired = false;

  @override
  Widget build(BuildContext context) {
    final value = _fired ? widget.pressedLabel : 'idle';
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        aid(
          widget.id,
          GestureDetector(
            onScaleUpdate: (details) {
              final recognized = widget.channel == _Channel.zoom ? details.scale != 1.0 : details.rotation != 0.0;
              if (recognized && !_fired) setState(() => _fired = true);
            },
            child: Container(
              width: 280,
              height: 120,
              alignment: Alignment.center,
              color: Theme.of(context).colorScheme.surfaceContainerHighest,
              child: Text(widget.label),
            ),
          ),
        ),
        aidValue('${widget.id}.value', value, Text(value)),
      ],
    );
  }
}
